### DISCUSSION TITLE
Why is the Popover modal by default?

### DISCUSSION THREAD

**Discussion by loicplaire** [adobe/react-spectrum#3802] [Q&A] [answered]
Hi!

First of all, I want to thank you for the amazing work you are doing! We are using `react-aria` in our components library and it helped us immensely so far!

I am currently building a Popover component using [usePopover](https://react-spectrum.adobe.com/react-aria/usePopover.html) and wondered why the popover is modal by default.
Most of the examples I have seen in other component libraries have non-modal Popover by default, i.e. you can scroll and interact with the page when a popover is opened:

- https://reakit.io/docs/popover/
- https://reshaped.so/content/docs/components/popover
- https://www.radix-ui.com/docs/primitives/components/popover
- and so on.

I can also see, that most of your pickers are modal (such as https://react-spectrum.adobe.com/react-spectrum/Picker.html) which seems a little bit unusual.

I can make the popover non-modal by using the `isNonModal` prop but I wonder why [you recommend against it in the props documentation](https://github.com/adobe/react-spectrum/blob/e4ab4895584b6f453157a78219b7b1e4a14adc06/packages/%40react-aria/overlays/src/usePopover.ts#L32-L40) when, from my understanding, it is very common for a popover to be non-modal. 

Would you be able to clarify this for me? Thank you so much!

**Comment by devongovett** [comment, marked as answer]
The reasons are mostly related to React portals.

1. Because overlays are rendered in a portal at the end of the document body, there may be a lot of elements between the original trigger element and the popover. This can make it very difficult for screen reader users (especially on mobile) to move between the trigger and the overlay. For this reason, we hide all elements outside the overlay from screen reader users which keeps their focus contained within the overlay.
2. As for scrolling, this was a recent change in the last release. The new behavior of preventing scrolling outside popovers/menus actually matches native behavior (at least on macOS). If you try a native `<select>` for example, you can see this. Previously we would close the popover when the surrounding element scrolled, which caused some usability issues (e.g. accidentally closing). Repositioning the popover so it always follows the trigger on scroll is also not possible in all cases (e.g. if the trigger is scrolled out of view, but the popover pops out of the scrolling container). It can also be quite janky to do this. The simplest solution to this is to prevent scrolling, which keeps the user's focus on the overlay.
3. Clicking outside popovers is also prevented so that when the user is trying to dismiss the popover, they don't accidentally activate something else. This also matches the behavior of a native `<select>` or menu on macOS.

> I can make the popover non-modal by using the isNonModal prop 

We strongly recommend against this, because it will cause problems for screen reader users as described in (1) above. ComboBox is an example of a component that specifically handles this case. It hides all elements except both the overlay _and_ the input, since the user's focus remains on the input while the popover is open. The user also needs to interact with the input (e.g. moving the cursor) while the menu is open, so we don't prevent clicking outside in this case. We are also forced to close the menu on scroll in this case to avoid the other issues described above.

Hopefully this clarifies the behavior. If you do have a case where scrolling or interaction outside a popover is expected, you can always drop down to the lower level hooks that `usePopover` is built with, and compose them together yourself, but proceed with caution.

**Reply by loicplaire** [reply]
Thanks @devongovett for the detailed answer, it makes sense! Outside of the accessibility benefits, I also like the fact that it also matches the native OS behaviours.

**Reply by sarimarton** [reply]
Chiming in from [twitter](https://mobile.twitter.com/sarimarton/status/1644134887262175232). I'd like to argue against the current status quo.

1. Screen readers as UX-impacting factor. I understand that accessibility is a top priority, it's even reflected in the name of the library, but it's legitimate to think of it as _priority within development effort_, and not as such that mainstream UX is sacrificed for accessibility reasons.

2. Fixed positioned triggers. A strong use-case for a non-modal popover is when it's in a fixed component (e.g. a header), and you want to scroll the content area, so pop-over scrolling is not an issue. An example is YouTube, where I'm checking out my notifications and the video feed simultaneously, and closing the notification feels like an unnecessary extra step, I wouldn't do otherwise, and it blocks me in navigating the content while keeping the notifications on screen as well.

3. macOS relevance. Although I could maybe even argue that macOS has some similarities with outside scrollability, because when you open a long page in Settings (e.g. the Keyboard section), and drop down a select box, you can still scroll the sidebar -- I'm actually hesitant in that following macOS should be a legitimate goal at all. Apple made a lot of questionable decisions lately in macOS, and basically no one is infallible. Also inherent differences between native/web might play a role in the question.

4. Accidental scroll. I wonder if accidental close by scrolling happened with scrollable dropdown content, where the mouse was at the edge of the dropdown? It might be an issue specific to scrollable dropdowns, and could be treated as that, because otherwise it doesn't sound reasonable to speculate on the accidental nature of user actions. In fact, I remember I developed a habit of nudging the scroll wheel up or down a bit outside _just to close the combobox_, because it takes less energy than clicking.

5. In complex, many-field forms quick discoverability is an important aspect, and having to click out every time in order to proceed with either scrolling or checking out another dropdown is a significant cost.

6. Combobox being treated as a special case is a smell. Walking between the field and the picker should be a default capability. useDatePicker also has an editable field, which is unreachable when the picker is open. It feels like a bug.

7. In the twitter thread I'm coming I mentioned content shift because of the scrollbar disappearance, whereas referring to edge cases was in the response. I wonder if it could be counted as an edge case as well, and if we add up the edge cases for both sides together, who comes out as a winner. Assuming that we give up popover scrolling. Just close the popover on outside scroll.

**Reply by devongovett** [reply]
Thanks for the feedback. To be clear, the point about screen readers is secondary to whether clicking outside takes two clicks or scrolling is disabled. It's something we did before we made those changes. Matching macOS was also an _additive_ reason we felt ok doing it, not a primary reason we did it. I would say the two main reasons were:

1. Accidental scrolling outside, or reaching the end of a scrollable region within the popover causing outside scrolling. This previously closed the popover and we had users reporting this as a bug/annoyance. Programmatic auto scrolling (e.g. via keyboard navigation) sometimes also caused page scrolling which closed the popover.
2. Accidentally activating other buttons outside the popover when just trying to close it, for example accidentally navigating to another page or performing an action by clicking another button. You had to be very careful where you clicked to close the popover to avoid this, which was another source of frustration and usability issues.

Fixed position overlays is an issue we are working on in general, but I could see still wanting to avoid accidental activation when trying to close. We could probably add some options so apps could decide based on their individual UX and what other elements are around the popover whether to allow scrolling/outside interactions, but I do think this is a safe default that avoids a lot of edge cases in general.

That said, [usePopover](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/overlays/src/usePopover.ts) is pretty simple and just composes a few other hooks together so it would be pretty easy to copy paste it and modify to suit your needs in the meantime.

**Reply by kruzliak-juraj** [reply]
Hi [devongovett](https://github.com/devongovett), I would like to add my inconsequential voice to this topic and agree with sarimarton reasoning. At the moment I am also building a design system and everything about react-aria seemed great so far until I came across this same issue and have the same or very similar reasons why it is not great as loicplaire and some others that I found in different issues across the past few years. While it might not be showstopper, it is a very big reconsider factor, even though I've already put a lot of work into implementing react-aria hooks. 

As OC post mentions, Radix, one of the biggest, have also different popover behavior by default.

Libraries that as far I as I know pride themselves to be built on top of yours like NextUI do not follow this pattern and 'hacked' their Select component to basically use `isNonModal`, and they also disable the hiding of the popover when scrolled (presumably taking the "so it would be pretty easy to copy and paste it and modify to suit your needs in the meantime." approach). For the Dropdown component, they also somehow managed to keep the scrollbar shown, so there is no content shift (but there is one in their Modal component). I could give examples from other libraries as well, but I think this one is most relevant since it uses react-aria as its bases.

It would be amazing if you and the team could take a look at this behavior once again and maybe reconsider.
Thank you for all the work so far.

**Reply by devongovett** [reply]
Is this just an opinion or does this cause an actual problem for your application? And which behavior in particular are you talking about?

**Reply by kruzliak-juraj** [reply]
It is an actual problem in terms of user experience, as we agreed within the team. We need to either lock out the user from scrolling and interacting with outside elements when popover is opened (in select, menu or others) -- or use combobox pattern everywhere that closes the popover when the user scrolls the page, which is also problematic from our user experience perspective. 
Some examples may include having a longer list in popover that also has some relevance to some other data that is beyond the fold, lower in the page, that the user would like to scroll to, and then they have to open the menu again since it closes after scrolling to it. In case of where it is modal, scrolling is not possible at all and user has to first close the menu which may be very confusing, as for why they can't suddenly scroll. 

Some other references for other people experiencing similar or same complications: 
- https://github.com/adobe/react-spectrum/issues/4213#issuecomment-1476175517 - you recommended going to the lower level hooks, same as in here, which is a ok temporary solution, but again I would like to see a change if it would be reasonable for you
- https://github.com/adobe/react-spectrum/issues/900#issuecomment-673285274 - older but still relevant somewhat

____

Sorry for confusing the topic with the Modal example I mentioned, about scrollbar content shift, that is also a small issue but not necessarily related to this.

**Reply by devongovett** [reply]
The problem is that you have to choose between preventing scrolling while popovers are open, or closing popovers when scrolling outside. Keeping popovers open while scrolling is not possible (ie what if the trigger they are attached to goes out of the viewport?). We decided that preventing scroll was better than closing on scroll (too easy to accidentally scroll).

**Reply by sarimarton** [reply]
@devongovett I remember you mentioned edge cases and hard implementation, but react-laag handles this case. Also, couldn't the new CSS Anchor Positioning provide a bug-free implementation?

**Reply by kruzliak-juraj** [reply]
I understand the worry of accidental scroll, yet it is a need in certain use cases, which is why I think should be covered. 

Regarding it not being possible, I am not sure what do you mean by that. The examples in the original post show it is possible, the trigger and the popover both just scroll away on scroll, while being still opened when the user scrolls back in the view. The same behavior as in the original examples can be found in [MUI poper](https://mui.com/material-ui/react-popper/), or [NextUI popover](https://nextui.org/docs/components/popover#colors), or [Mantine popover](https://mantine.dev/core/popover/), and I could go on. Besides, I would think (I may be wrong) that some kind of detection of if it is out of viewport could be done that would close it in such case, but I do not really see the need for that in my opinion. The feature of it behaving like in all the examples is satisfactory. 

Oddly enough, I just realized that I want the same behavior as your `Tooltip` has when it is controlled, as can be seen in this quick mock: https://codesandbox.io/p/sandbox/elegant-tom-w54n64 - so I guess this gives me another option on how to implement it. It is odd that it works there, but I guess you did not expect it to be used in such a way, which makes me think it would be risky to use it this way in case you change it there as well.

As I said, I just want to plea for you to reconsider it and/or give it as an additional behavior option. If you want to keep the current implementation, I will understand, but we will have to either create a more complex solution (or use `useTooltip` api :)) on our end or reconsider using this library.

Edit: after more testing, it seems that controlled `Popover` works the same as the `Tooltip` so I guess I will play around with that for now and see if it creates some issues somewhere else that I am not aware of.

**Reply by devongovett** [reply]
It's not an edge case. One of the main reasons to use a popover is that it pops "out" of whatever container might be clipping it, be that a dialog, scrolling list, etc. Like this:
<img width="400" alt="image" src="https://github.com/adobe/react-spectrum/assets/19409/fa313edc-914f-4380-906d-d16eed2f9eae">

If the user were to scroll that container such that the trigger is out of view, there's no longer anything for the popover to attach to: 
<img width="400" alt="image" src="https://github.com/adobe/react-spectrum/assets/19409/50f66c99-58e9-4279-bb46-1f43862a3836">

This problem exists no matter what library you use. Most of the examples you pointed to work because the trigger element and the popover are both inside the same scrollable area (the body), so when you scroll the trigger out of view the popover also goes out of view. If the popover pops out of a nested scrollable area, then you will see this problem. The solutions are either to hide the popover on scroll or prevent scrolling. We think the latter is a better UX.

**Reply by kruzliak-juraj** [reply]
Thank you for the explanation and the time, that makes sense. So it is ok only if I could guarantee that both elements will be the body scrollable area. Thank you very much, I have much to discuss with the team :)

**Reply by kruzliak-juraj** [reply]
One last quick note, after testing scroll area in scroll area with popover in different libraries, it seems they opted in for re-calculating the position on scroll.
Radix example - https://codesandbox.io/p/sandbox/radix-primitives-tailwind-forked-xtpngy
MUI example - https://codesandbox.io/p/sandbox/mui-5-forked-g3hdfl
Mantine example - https://codesandbox.io/p/sandbox/mantine-v7-forked-fvzntk

**Comment by borispetrovdev** [comment]
I’d also think scrolling behind a popover should be par for the course, and I can’t help but point out the irony of the GitHub comments interface itself exhibiting this behavior: 

https://github.com/user-attachments/assets/5d5ba6e2-d053-45a4-8c47-4cf0807a4e3f

That said, I appreciate that the React Aria team is grappling with genuine tradeoffs, and that the option of dropping down to the lower level hooks is reasonable.
